vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/config.go

```diff
@@ -423,9 +437,32 @@ func (c *Config) Complete(informers informers.SharedInformerFactory) CompletedCo
 // Complete fills in any fields not set that are required to have valid data and can be derived
 // from other fields. If you're going to `ApplyOptions`, do that first. It's mutating the receiver.
 func (c *RecommendedConfig) Complete() CompletedConfig {
+	if c.ClientConfig != nil {
+		if ns := eventRefNamespace(); len(ns) == 0 {
+			klog.V(2).Infof("Neither running in-cluster, nor is POD_NAMESPACE set. Skipping life-cycle events.")
+		} else {
+			c.LifecycleEventSink = &v1.EventSinkImpl{
+				Interface: kubernetes.NewForConfigOrDie(c.ClientConfig).CoreV1().Events(ns),
+			}
+		}
+	}
+
 	return c.Config.Complete(c.SharedInformerFactory)
 }
 
+func eventRefNamespace() string {
+	ns := os.Getenv("POD_NAMESPACE")
+
+	if len(ns) == 0 {
+		bs, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
+		if err == nil {
```

Comment:
can we handle the error case as well? i guess this will return empty namespace?

Reply:
fixed